Add RFC: Providing fallback credentials on timeout (#2218)
awslabs/smithy-rs: design/src/rfcs/rfc0031_providing_fallback_credentials_on_timeout.md

RFC: Providing fallback credentials on external timeout
=======================================================

> Status: RFC
>
> Applies to: client

For a summarized list of proposed changes, see the [Changes Checklist](#changes-checklist) section.

> This RFC proposes a fallback mechanism for credentials providers on external timeout (see the [Terminology](#terminology) section), allowing them to continue serving (possibly stale) credentials for the sake of overall reliability of the intended service; the IMDS credentials provider is an example that must fulfill such a requirement to support static stability.

Reviewer: As someone who is an unrepentant semicolon addict, I must note that the first word after a semicolon should be capitalized 😝 .
```suggestion
This RFC proposes a fallback mechanism for credentials providers on external timeout (see the [Terminology](#terminology) section), allowing them to continue serving (possibly stale) credentials for the sake of overall reliability of the intended service; The IMDS credentials provider is an example that must fulfill such a requirement to support static stability.
```

Terminology
-----------

> - External timeout: A timeout within which a call to the `provide_credentials` method must complete. `provide_credentials` returns a future and if it does not complete within this timeout, it gets dropped on the floor and there is no output retrieved from the future.

Reviewer: This sentence felt a bit wordy to me so I tried my hand at re-wording it. Let me know if you think it makes anything clearer.

```suggestion
- External timeout: The name of the timeout that occurs when a duration elapses before an async call to `provide_credentials` returns. In this case, `provide_credentials` returns no credentials.
```

> - Internal timeout: A timeout imposed on functions/methods internally called by `provide_credentials` that, when failing to meet any of them, renders `provide_credentials` return an `CredentialsError`. For instance, internal timeout includes connection timeout, TLS negotiation timeout, and HTTP request timeout, if `provide_credentials` needs to fetch credentials via HTTP(S). This is a kind of timeout `provide_credentials` can react and handle it accordingly, i.e. returning fallback credentials or an `CredentialsError`.

Reviewer: I had a go at this one too
```suggestion
- Internal timeout: The name of the timeout that occurs when a duration elapses before an async call to some function, inside the implementation of `provide_credentials`, returns. Examples include connection timeouts, TLS negotiation timeouts, and HTTP request timeouts. Implementations of `provide_credentials` may handle these failures at their own discretion e.g. by returning _(possibly expired)_ credentials or a `CredentialsError`.
```

Assumption
----------

> This RFC is concerned only with external timeout, as the cost of designing APIs poorly is much higher than that for internal timeout. The former will affect a public trait implemented by all credentials providers whereas the latter can be handled locally by individual credentials providers without affecting one another.

Reviewer: ```suggestion
This RFC is concerned only with external timeouts, as the cost of poor API design is much higher than in this case than for internal timeouts. The former will affect a public trait implemented by all credentials providers whereas the latter can be handled locally by individual credentials providers without affecting one another.
```

Problem
-------

> We have mentioned static stability. Supporting it calls for the following functional requirement, among others:

Reviewer: "static stability" should be included in the terminology section

Author: A good suggestion. Addressed in d413e58.

- REQ 1: Once a credentials provider has served credentials, it should continue serving them in the event of a timeout (whether internal or external) while obtaining refreshed credentials.

Today, we have the following trait method to obtain credentials:
```rust
fn provide_credentials<'a>(&'a self) -> future::ProvideCredentials<'a>
where
    Self: 'a,
```

> This method can be used in async contexts, which means that nothing stops it from being raced against a timeout future, as demonstrated by the following code snippet from `LazyCredentialsCache`:

Reviewer: ```suggestion
This method returns a future, which can be raced against a timeout future as demonstrated by the following code snippet from `LazyCredentialsCache`:
```

```rust
let timeout_future = self.sleeper.sleep(self.load_timeout); // by default self.load_timeout is 5 seconds.
// --snip--
let future = Timeout::new(provider.provide_credentials(), timeout_future);
let result = cache

>    .get_or_load(|| {
>         async move {
>             let credentials = future.await.map_err(|_err| {
>                 CredentialsError::provider_timed_out(load_timeout)
>             })??;
>             // --snip--
>         }
>     }).await;

Reviewer: I think this shorter version is also valid
```suggestion
   .get_or_load(|| async move {
        let credentials = future.await.map_err(|_err| {
            CredentialsError::provider_timed_out(load_timeout)
        })??;
        // --snip--
    }).await;
```

</p>

The figure above illustrates an example. This `CredentialsProviderChain` consists of three credentials providers. When `CredentialsProviderChain::provide_credentials` is called, provider 1's `provide_credentials` is called but does not find credentials so passes the torch to provider 2, which in turn successfully loads credentials and returns them. The next time the method is called, provider 1 does not find credentials but neither does provider 2 this time, because an external timeout by `timeout_future` given to the whole chain kicked in and the future is dropped while provider 2's `provide_credentials` was running. Given the functional requirements, provider 2 should return the previously available credentials but today the code snippet from `LazyCredentialsCache` returns a `CredentialsError::ProviderTimedOut` instead.

**Case 2:** Provider 2 successfully loaded credentials but later was not reached because its preceding provider was still running when an external timeout kicked in.

<p align="center">
<img width="750" alt="chain-provider-ext-timeout-2" src="https://user-images.githubusercontent.com/15333866/212421712-8c6eab11-a0c1-4229-8ba3-67b0bb6056e7.png">
</p>

> The figure above illustrates an example with the same setting as the previous figure. Again, when `CredentialsProviderChain::provide_credentials` is called the first time, provider 1 does not find credentials but provider 2 does. The next time the method is called, provider 1 is still executing `provide_credentials` and then an external time by `timeout_future` kicked in. Consequently, the execution of `CredentialsProviderChain::provide_credentials` has been terminated. Given the functional requirements, provider 2 should return the previously available credentials but today the code snippet from `LazyCredentialsCache` returns `CredentialsError::ProviderTimedOut` instead.

Reviewer: ```suggestion
The figure above illustrates an example with the same setting as the previous figure. Again, when `CredentialsProviderChain::provide_credentials` is called the first time, provider 1 does not find credentials but provider 2 does. The next time the method is called, provider 1 is still executing `provide_credentials` and then an external timeout by `timeout_future` kicked in. Consequently, the execution of `CredentialsProviderChain::provide_credentials` has been terminated. Given the functional requirements, provider 2 should return the previously available credentials but today the code snippet from `LazyCredentialsCache` returns `CredentialsError::ProviderTimedOut` instead.
```

Proposal
--------
To address the problem in the previous section, we propose to add a new method to the `ProvideCredentials` trait called `on_timeout` that looks something like this:
```rust
mod future {
    // --snip--

>     pub struct OnTimeout<'a>(NowOrLater<Option<Credentials>, BoxFuture<'a, Option<Credentials>>>);

Reviewer: related, possibly useful: https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.now_or_never

Author: Hmm, it may help obviate the need for a potential timeout on `on_timeout` as [John asked](https://github.com/awslabs/smithy-rs/pull/2218#discussion_r1072657848). We can have `on_method` as `async` and yet sort of behave synchronously; It returns `Some` from `Poll::Ready` but `None` from `Poll::Pending`.

Author: Interesting. It may help obviate the need for a potential timeout on `on_timeout` as [John asked](https://github.com/awslabs/smithy-rs/pull/2218#discussion_r1072657848). We can have `on_method` as `async` and yet sort of behave synchronously; The future will not wonder off and get polled indefinitely.

Author: Addressed in 413e58.

Author: Addressed in d413e58.

    // impls for OnTimeout similar to those for the ProvideCredentials future newtype
}

pub trait ProvideCredentials: Send + Sync + std::fmt::Debug {
    // --snip--

>     fn on_timeout<'a>(&'a self) -> future::OnTimeout<'a> {

Reviewer 2: Is there any reason this shouldn't be synchronous? What's the use-case for asynchronously retrieving the credentials on timeout? Would we have a timeout for the `on_timeout`?

Author: No reason it shouldn't. I think each has pros/cons:
- synchronous
**pros**: Easier to reason about. 
**cons**: Will be a breaking change if we want to switch to async in the future.
- asynchronous
**pros**: It's async from the beginning, so less likely to introduce a breaking change.
**cons**: Exactly what you brought up, may need to specify a timeout.

Currently, there are no concrete use cases requiring this to be async.

Author: An async primitive is more suitable as discussed (since we do not have an async critical section). The method is a fallback plan so the execution should not take too long. d413e58 mimics `FutureExt::now_or_never` using our `NowOrLater` with `OnlyReady`, which means no need to have yet another timeout for the method.

Author: > Would we have a timeout for the on_timeout?

It is possible, although I don't think we specify that as an input argument to `fallback_on_interrupt` for the same reason I described in the Alternative section in the RFC. Whoever uses `fallback_on_timeout` can race it against a timeout future just like `LazyCredentialsCache` does with `provide_credentials` today. It just that the timeout for `fallback_on_timeout` will probably way shorter than that for `provide_credentials`, as the execution of `fallback_on_timeout` shouldn't take too long.

Haven't decided whether the timeout for `fallback_on_timeout` should be introduced today (something like `load_timeout` for `LazyCredentialsCache`). If we are to introduce it later, we need to see if we can do so without causing breaking changes.

Author: > Would we have a timeout for the on_timeout?

It is possible, although I don't think we specify that as an input argument to `fallback_on_interrupt` for the same reason I described in the Alternative section in the RFC. Whoever uses `fallback_on_timeout` can race it against a timeout future just like `LazyCredentialsCache` does with `provide_credentials` today. It just that the timeout for `fallback_on_timeout` will probably be way shorter than that for `provide_credentials`, as the execution of `fallback_on_timeout` shouldn't take too long.

Haven't decided whether the timeout for `fallback_on_timeout` should be introduced today (something like `load_timeout` for `LazyCredentialsCache`). If we are to introduce it later, we need to see if we can do so without causing breaking changes.

        future::OnTimeout::ready(None)
    }
}
```
This method allows credentials providers to have a fallback mechanism on an external timeout and to serve credentials to users if needed. It comes with a default implementation to return a future that immediately yields `None`. Credentials providers that need a different behavior can choose to override the method.

The user experience for the code snippet in question will look like this once this proposal is implemented:
```rust
let timeout_future = self.sleeper.sleep(self.load_timeout); // by default self.load_timeout is 5 seconds.
// --snip--
let future = Timeout::new(provider.provide_credentials(), timeout_future);
let result = cache
    .get_or_load(|| {
        async move {

>            let credentials= match future.await {

Reviewer 2: ```suggestion
           let credentials = match future.await {
```

                Ok(creds) => creds?,
                Err(_err) => match provider.on_timeout().await { // can provide fallback credentials
                    Some(creds) => creds,
                    None => return Err(CredentialsError::provider_timed_out(load_timeout)),
                }
            };
            // --snip--
        }
    }).await;
// --snip--
```

How to actually implement this RFC
----------------------------------

Almost all credentials providers do not have to implement their own `on_timeout` except for `CredentialsProviderChain` (`ImdsCredentialsProvider` also needs to implement `on_timeout` when we are adding static stability support to it but that is outside the scope of this RFC).

> Considering the two cases we analyzed above, implementing `CredentialsProviderChain::on_timeout` is not so straightforward. Keeping track of whose turn in the chain it is to call `provide_credentials` when an external timeout has ocurred is a challening task. Even if we figured it out, that would still not satisfy `Case 2` above, because it was provider 1 that was actively running when the external timeout kicked in, but the chain should return credentials from provider 2, not from provider 1.

Reviewer: ```suggestion
Considering the two cases we analyzed above, implementing `CredentialsProviderChain::on_timeout` is not so straightforward. Keeping track of whose turn in the chain it is to call `provide_credentials` when an external timeout has occurred is a challenging task. Even if we figured it out, that would still not satisfy `Case 2` above, because it was provider 1 that was actively running when the external timeout kicked in, but the chain should return credentials from provider 2, not from provider 1.
```

> With all that taken into account, we can consider starting with the following simple approach:

Reviewer: ```suggestion
With that in mind, consider instead the following simple approach:
```

```rust
impl ProvideCredentials for CredentialsProviderChain {
    // --snip--

    fn on_timeout<'a>(&'a self) -> future::OnTimeout<'a> {
        future::OnTimeout::new(async move {
            for (_, provider) in &self.providers {
                match provider.on_timeout().await {
                    creds @ Some(_) => return creds,
                    None => {}
                }
            }
            None
        })
    }
}
```

> `CredentialsProviderChain::on_timeout` will invoke each provider's `on_timeout` method until credentials are returned by one of them. It ensures that the updated code snippet for `LazyCredentialsCache` can return credentials from provider 2 in both `Case 1` and `Case 2`. Even if `timeout_future` wins the race, the execution subsequently calls `provider.on_timeout().await` to obtain fallback credentials from provider 2, assuming provider 2's `on_timeout` is implemented to return credentials retrieved on the first call to `CredentialsProviderChain::provide_credentials`.

Reviewer: > the execution subsequently calls `provider.on_timeout().await` to obtain fallback
this is a possible place to use `now_or_never`

Author: I believe we discussed this offline? The execution does need to call `fallback_on_interrupt` to obtain fallback credentials because the future for `provide_credentials` gets dropped and we cannot do anything with it. The concept of `now_or_never` does help in implementing the future returned by `fallback_on_interrupt`.

The downside of this simple approach is that the behavior is not clear if more than one credentials provider in the chain can return credentials from their `on_timeout`. Note, however, that it is the exception rather than the norm for a provider's `on_timeout` to return fallback credentials, at least at the time of writing (01/13/2023). The fact that it returns fallback credentials means that the provider successfully loaded credentials at least once, and it usually continues serving credentials on subsequent calls to `provide_credentials`.

Should we have more than one provider in the chain that can potentially return fallback credentials from `on_timeout`, we could configure the behavior of `CredentialsProviderChain` managing in what order and how each `on_timeout` should be executed; it can be possible to introduce a new API without breaking changes. That being said, we first need to investigate and understand the use case behind it to design the right API.

Alternative
-----------

> In this section, we will describe our alternative approach that we ended up dismissing as unworkable.

Reviewer: ```suggestion
In this section, we will describe an alternative approach that we ended up dismissing as unworkable.
```

                .provide_credentials_with_timeout(Arc::clone(&sleeper), /* how do we calculate timeout for each provider ? */)
                .await
            {
                Ok(credentials) => {
                    return Ok(credentials);
                }
                Err(CredentialsError::ProviderTimedOut(_)) => {
                    // --snip--
                }
                Err(err) => {
                   // --snip--
                }
           }
        }
        Err(CredentialsError::provider_timed_out(timeout))
    }
```
There are mainly two problems with this approach. The first problem is that as shown above, there is no sensible way to calculate a timeout for each provider in the chain. The second problem is that exposing a parameter like `timeout` at a public trait's level is giving too much control to users; delegating overall timeout to the individual provider means each provider has to get it right.

Changes checklist
-----------------

- [ ] Add `on_timeout` method to the `ProvideCredentials` trait with the default implementation
- [ ] Implement the `OnTimeout` future newtype
- [ ] Implement `CredentialsProviderChain::on_timeout`

> - [ ] Add unit tests for `Case 1` and `Case 2`

Reviewer: When new credentials come through, we should test that all layer's cached credentials are refreshed.

Author: Thanks for coming up with the use case! As discussed, we can employ a different approach. See the [Appendix](https://github.com/awslabs/smithy-rs/blob/d413e58e57dc0fd7bcd68583414bb7aa250c684c/design/src/rfcs/rfc0031_providing_fallback_credentials_on_timeout.md#appendix) section.
